[ML] Adds ES|QL support for field statistics table in Discover

qn895 opened the pull request:
## Summary

Follow up of https://github.com/elastic/kibana/pull/179098. Adds ES|QL for support field statistics table in Discover 

<img width="1728" alt="Screenshot 2024-05-09 at 13 59 29" src="https://github.com/elastic/kibana/assets/43350163/ede8bcc5-2fc6-4a0a-8e3e-993294bbeafd">

https://github.com/elastic/kibana/assets/43350163/09a1feeb-cff6-4b8b-bf82-bb930413f215


https://github.com/elastic/kibana/assets/43350163/2acd78ec-1856-406e-a8f9-f6b2a26f8d81



### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

qn895 committed:
Utilize include_empty_fields in fieldCaps to reduce data visualizer requests

qn895 committed:
Fix set component

qn895 committed:
Fix duplicated getFieldsForWildcard

qn895 committed:
Clean up

qn895 committed:
Make visible field names from discover also include multi fields

qn895 committed:
Clean up & fix types

qn895 committed:
Fix tests, fix runtime fields

qn895 committed:
Fix count for non aggregetable fields

qn895 committed:
Merge remote-tracking branch 'upstream/main' into dv-no-empty-fields

qn895 committed:
Fix empty field changes

qn895 committed:
Add ES|QL support for discover

qn895 committed:
Fix ES|QL id

qn895 committed:
[ES|QL] Test dropNullColumns

qn895 committed:
Merge remote-tracking branch 'upstream/main' into dv-esql-empty-fields

qn895 committed:
Add default limit

qn895 committed:
Merge remote-tracking branch 'upstream/main' into dv-esql-empty-fields

qn895 committed:
Use initial fetch call

qn895 committed:
Merge remote-tracking branch 'upstream/main' into dv-esql-empty-fields

qn895 committed:
Clean up comments

qn895 committed:
Clean up

qn895 committed:
Merge remote-tracking branch 'upstream/main' into dv-esql-empty-fields

qn895 committed:
Fix templates

qn895 committed:
Fix locations

qn895 committed:
Git reset  .buildkite/pipeline-resource-definitions/fix-location-collection.ts

qn895 committed:
Fix geo timestamp, fix latest version

qn895 committed:
Make sample size messaging match with unifiedfieldlist popup

qn895 committed:
Add new top values equivalent to show % in rows, % in all flattened values

stratoula committed:
Merge branch 'main' into dv-esql-empty-fields

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

qn895 committed:
Revert "[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'"

This reverts commit 684cb9d208f01d30edf2f068d967cb5ce66672a7.

qn895 committed:
Revert changes in Discover

qn895 committed:
Revert "Revert changes in Discover"

This reverts commit 0e0df6fd432f5b5241b677fd397db7f7078e02d5.

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

elasticmachine commented:
Pinging @elastic/kibana-esql (Team:ESQL)

qn895 commented:
/ci

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

stratoula commented:
@qn895 this looks great. Which is the difference between top values and top sampled values?

<img width="1113" alt="image" src="https://github.com/elastic/kibana/assets/17003240/178821a1-8087-4963-bee7-a6006186e60d">

elasticmachine commented:
Pinging @elastic/obs-ux-infra_services-team (Team:obs-ux-infra_services)

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ml-field-stats-esql-discover

qn895 committed:
Merge relevant changes

qn895 committed:
Move setQueryHistoryStatus up top

qn895 committed:
Remove fetch for sampled values

qn895 committed:
Update functional test to assert view mode toggle even in es|ql

qn895 committed:
Fix Discover tests

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ml-field-stats-esql-discover

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

qn895 committed:
Fix backtick in test

qn895 committed:
Update group4 test

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ml-field-stats-esql-discover

qn895 committed:
Fix unused var

qn895 committed:
Update i18n

qn895 committed:
Change wording to 'estimated' for esql keywords

qn895 committed:
Update type

peteharverson commented on x-pack/plugins/data_visualizer/public/application/common/components/top_values/top_values.tsx:
```diff
@@ -69,7 +71,7 @@ export const TopValues: FC<Props> = ({
 
   if (stats === undefined || !stats.topValues) return null;
-  const { fieldName, sampleCount } = stats;
+  const { fieldName, sampleCount, approximate } = stats;
```
Did you decide not to show an 'Other' row for cases like this?

<img width="681" alt="Screenshot 2024-05-14 at 11 28 58" src="https://github.com/elastic/kibana/assets/7405507/6729d276-2d17-4ea3-80e5-07c47ca0e451">

qn895 replied:
Yes for now `Other` would be hard to calculate for ES|QL currently doesn't support tracking total hits. And we cannot base it off the cardinality/count because the requests are randomly sampled, and therefore can change.

peteharverson reviewed (commented):
Tested (inside Discover and adding a panel to a dashboard) and overall looks good. Just left one question around the top values.

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ml-field-stats-esql-discover

qn895 committed:
Update top values

peteharverson reviewed (approved):
LGTM

qn895 committed:
Update tests

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ml-field-stats-esql-discover

qn895 committed:
Update show view mode toggle setting

qn895 committed:
Add proxied services

stratoula committed:
Merge branch 'main' into ml-field-stats-esql-discover

stratoula commented:
/ci

stratoula commented on src/plugins/discover/public/application/main/components/field_stats_table/types.ts:
```diff
@@ -164,4 +164,9 @@ export interface FieldStatisticsTableProps {
    */
   searchSessionId?: string;
+  /**
+   * If table should should accept text based query
```
```suggestion
   * If table should accept text based query
```

stratoula commented on src/plugins/discover/public/application/main/components/field_stats_table/types.ts:
```diff
@@ -164,4 +164,9 @@ export interface FieldStatisticsTableProps {
+   * If table should should accept text based query
+   * isPlainRecord is true when in ES|QL mode
+   */
+  isPlainRecord?: boolean;
```
@davismcphee here https://github.com/elastic/kibana/pull/183108 is changing the `isPlainRecord` to `isEsqlMode `. Let's be proactive and use `isEsqlMode` here instead

stratoula reviewed (commented):
Thanx @qn895 this looks great and is a fantastic addition for the ES|QL mode. I have some questions mostly because is the first time I take a look in this functionality.

1. I see that we are running a very big query for getting some data. Do we really need at the first page to get all metrics for all fields? why don't we only get the metrics for the fields we are showing in this page?  It will make this query faster. But I might miss something

<img width="1663" alt="image" src="https://github.com/elastic/kibana/assets/17003240/e5288ac6-2a69-4e91-bd39-f2f9683a3ab1">

2. I see that we are using the synchronous strategy, can we use the async one instead? It will be more performant

jughosta commented on src/plugins/discover/public/application/main/utils/get_valid_view_mode.ts:
```diff
@@ -20,10 +20,5 @@ export const getValidViewMode = ({
-    // only this mode is supported for text-based languages
-    return VIEW_MODE.DOCUMENT_LEVEL;
-  }
-
```
@stratoula Do we want to enable now adding ES|QL field stats as a panel to dashboards? 

In this PR it resets `viewMode` to `documents` still when saving a search. Maybe because of https://github.com/elastic/kibana/blob/828b5c9da4c6c17b119d6e1c2048254e1b4cb779/src/plugins/discover/public/application/main/state_management/utils/build_state_subscribe.ts#L55

stratoula replied:
Does this line enable that? Hmm, is it possible on dataview mode?

stratoula replied:
I see, I didn't know that we have this functionality 😄 It is not very obvious that when you save a saved search in the field statistics tab, you embed this view instead of the table. 🤔 

My personal opinion is for now to always embed the datatable and not store the selected tab. I will add it as a discussion point at the next ES|QL meeting to decide the next steps. Maybe we can improve the experience here.

jughosta replied:
I just tested it again in this PR and looks like it got changed to show Field Statistics as a panel if the tab was active when saving in ES|QL mode too. I don't mind it either way but would be great to agree now how it should work.

stratoula replied:
I find it a very weird behavior tbh. I have added this as a conversation item for next week but I dont want to block this PR. Can we disable this behavior only for ESQL for now and create a follow up PR if necessary to enable it after the bi-weekly?

jughosta reviewed (commented):
Discover changes overall look good to me! Thanks! Left a couple of questions regarding services and that `viewMode` gets reset on save.

Also we need to be careful when resolving conflicts with https://github.com/elastic/kibana/pull/178916

stratoula commented on src/plugins/discover/public/application/main/components/field_stats_table/field_stats_table.tsx:
```diff
@@ -46,6 +47,14 @@ export const FieldStatisticsTable = (props: FieldStatisticsTableProps) => {
 
   const services = useDiscoverServices();
+
+  // Other apps consuming Discover UI might inject their own proxied data services
+  // so we need override the kibana context services with the injected proxied services
+  // to make sure the table use the right service
```
Yes this is a good question, I was also confused with that.

qn895 replied:
I've seen this happen in the [Logs explorer page in Observability serverless where the data's plugin timefilter service on that page is not the same default one](https://github.com/elastic/kibana/blame/main/x-pack/plugins/observability_solution/logs_explorer/public/controller/create_controller.ts/#L56-L57) as the one we were consuming (from the data visualizer plugin's side). This was causing some inconsistencies. I think this override is a reasonable one cause if there's any custom services provided by the solutions, we should be consuming them anyway.

qn895 committed:
Change to async esql strategy

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ml-field-stats-esql-discover

qn895 committed:
Condense mv_min count

qn895 committed:
Merge upstream/main into branch

qn895 committed:
Fix broken view mode update after merge

stratoula committed:
Merge branch 'main' into ml-field-stats-esql-discover

stratoula commented:
This looks great @qn895!! Thanx for addressing all the comments. 

I see there is one FT failing and also I see we are running this query

```
from logstash-* | LIMIT 10000| limit 10| STATS `machine.ram_count` = COUNT(MV_MIN(`machine.ram`)), `machine.ram_cardinality` = COUNT_DISTINCT(`machine.ram`),`referer_count` = COUNT(MV_MIN(`referer`)), `referer_cardinality` = COUNT_DISTINCT(`referer`),`relatedContent.article:modified_time_count` = COUNT(MV_MIN(`relatedContent.article:modified_time`)), `relatedContent.article:modified_time_cardinality` = COUNT_DISTINCT(`relatedContent.article:modified_time`),`relatedContent.article:published_time_count` = COUNT(MV_MIN(`relatedContent.article:published_time`)), `relatedContent.article:published_time_cardinality` = COUNT_DISTINCT(`relatedContent.article:published_time`),`relatedContent.article:section_count` = COUNT(MV_MIN(`relatedContent.article:section`)), `relatedContent.article:section_cardinality` = COUNT_DISTINCT(`relatedContent.article:section`),`relatedContent.article:section.keyword_count` = COUNT(MV_MIN(`relatedContent.article:section.keyword`)), `relatedContent.article:section.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.article:section.keyword`),`relatedContent.article:tag_count` = COUNT(MV_MIN(`relatedContent.article:tag`)), `relatedContent.article:tag_cardinality` = COUNT_DISTINCT(`relatedContent.article:tag`),`relatedContent.article:tag.keyword_count` = COUNT(MV_MIN(`relatedContent.article:tag.keyword`)), `relatedContent.article:tag.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.article:tag.keyword`),`relatedContent.og:description_count` = COUNT(MV_MIN(`relatedContent.og:description`)), `relatedContent.og:description_cardinality` = COUNT_DISTINCT(`relatedContent.og:description`),`relatedContent.og:description.keyword_count` = COUNT(MV_MIN(`relatedContent.og:description.keyword`)), `relatedContent.og:description.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.og:description.keyword`),`relatedContent.og:image_count` = COUNT(MV_MIN(`relatedContent.og:image`)), `relatedContent.og:image_cardinality` = COUNT_DISTINCT(`relatedContent.og:image`),`relatedContent.og:image.keyword_count` = COUNT(MV_MIN(`relatedContent.og:image.keyword`)), `relatedContent.og:image.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.og:image.keyword`),`relatedContent.og:image:height_count` = COUNT(MV_MIN(`relatedContent.og:image:height`)), `relatedContent.og:image:height_cardinality` = COUNT_DISTINCT(`relatedContent.og:image:height`),`relatedContent.og:image:height.keyword_count` = COUNT(MV_MIN(`relatedContent.og:image:height.keyword`)), `relatedContent.og:image:height.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.og:image:height.keyword`),`relatedContent.og:image:width_count` = COUNT(MV_MIN(`relatedContent.og:image:width`)), `relatedContent.og:image:width_cardinality` = COUNT_DISTINCT(`relatedContent.og:image:width`),`relatedContent.og:image:width.keyword_count` = COUNT(MV_MIN(`relatedContent.og:image:width.keyword`)), `relatedContent.og:image:width.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.og:image:width.keyword`),`relatedContent.og:site_name_count` = COUNT(MV_MIN(`relatedContent.og:site_name`)), `relatedContent.og:site_name_cardinality` = COUNT_DISTINCT(`relatedContent.og:site_name`),`relatedContent.og:site_name.keyword_count` = COUNT(MV_MIN(`relatedContent.og:site_name.keyword`)), `relatedContent.og:site_name.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.og:site_name.keyword`),`relatedContent.og:title_count` = COUNT(MV_MIN(`relatedContent.og:title`)), `relatedContent.og:title_cardinality` = COUNT_DISTINCT(`relatedContent.og:title`),`relatedContent.og:title.keyword_count` = COUNT(MV_MIN(`relatedContent.og:title.keyword`)), `relatedContent.og:title.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.og:title.keyword`),`relatedContent.og:type_count` = COUNT(MV_MIN(`relatedContent.og:type`)), `relatedContent.og:type_cardinality` = COUNT_DISTINCT(`relatedContent.og:type`),`relatedContent.og:type.keyword_count` = COUNT(MV_MIN(`relatedContent.og:type.keyword`)), `relatedContent.og:type.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.og:type.keyword`),`relatedContent.og:url_count` = COUNT(MV_MIN(`relatedContent.og:url`)), `relatedContent.og:url_cardinality` = COUNT_DISTINCT(`relatedContent.og:url`),`relatedContent.og:url.keyword_count` = COUNT(MV_MIN(`relatedContent.og:url.keyword`)), `relatedContent.og:url.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.og:url.keyword`),`relatedContent.twitter:card_count` = COUNT(MV_MIN(`relatedContent.twitter:card`)), `relatedContent.twitter:card_cardinality` = COUNT_DISTINCT(`relatedContent.twitter:card`),`relatedContent.twitter:card.keyword_count` = COUNT(MV_MIN(`relatedContent.twitter:card.keyword`)), `relatedContent.twitter:card.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.twitter:card.keyword`),`relatedContent.twitter:description_count` = COUNT(MV_MIN(`relatedContent.twitter:description`)), `relatedContent.twitter:description_cardinality` = COUNT_DISTINCT(`relatedContent.twitter:description`),`relatedContent.twitter:description.keyword_count` = COUNT(MV_MIN(`relatedContent.twitter:description.keyword`)), `relatedContent.twitter:description.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.twitter:description.keyword`),`relatedContent.twitter:image_count` = COUNT(MV_MIN(`relatedContent.twitter:image`)), `relatedContent.twitter:image_cardinality` = COUNT_DISTINCT(`relatedContent.twitter:image`),`relatedContent.twitter:image.keyword_count` = COUNT(MV_MIN(`relatedContent.twitter:image.keyword`)), `relatedContent.twitter:image.keyword_cardinality` = COUNT_DISTINCT(`relatedContent.twitter:image.keyword`)
```

in the first page while we are not depicting these stats. I am expecting this query to run in the second page. I wonder if this is possible with our current architecture.

<img width="1680" alt="image" src="https://github.com/elastic/kibana/assets/17003240/56e6b748-5169-4883-a5ba-cdf05ee8db63">

qn895 committed:
Fix url change viewMode from undefined to document causing unsaved change badge to persist

qn895 commented:
> I see there is one FT failing and also I see we are running this query


@stratoula You're right. I thought this was an unrelated flaky test but turns out it was due to the way the appState.viewMode was being set. Changing viewMode from `undefined` to `documents` was causing it to see that as an unsaved change. I've fixed it [here](https://github.com/elastic/kibana/pull/180849/commits/ebc84c7f3fd14ca65741c4e5485f21e966429718). 

> ```
> from logstash-* | LIMIT 10000| limit 10| STATS `machine.ram_count` = COUNT(MV_MIN(`machine.ram`)), `machine.ram_cardinality` = COUNT_DISTINCT(`machine.ram`),`referer_count` = COUNT(MV_MIN(`referer`)), `referer_cardinality` = COUNT_DISTINCT(`referer`),`relatedContent.article:modified_time_count` = COUNT(MV_MIN(`relatedContent.article:modified_time`)), `relatedContent.article:modified_time_cardinality` = COUNT_DISTINCT(`relatedContent.article:modified_time`),`relatedContent.article:published_time_count` = COUNT(MV_MIN(`relatedContent.article:published_time`)), `relatedContent.article:published_time_cardinality` = COUNT_DISTINCT(`relatedContent.article:published_time`),`relatedContent.article:section_count` = COUNT(MV_MIN(`relatedContent.article:section`)), ...
> ```
> 
> in the first page while we are not depicting these stats. I am expecting this query to run in the second page. I wonder if this is possible with our current architecture.
> 

The above query here is only getting the count and distinct count for these fields, which is needed for sorting on this table:

https://github.com/elastic/kibana/assets/43350163/a60899c0-3d1d-4bb2-b8ee-f307015f94a4

If we choose to disable sorting on the **Documents (%)** and **Distinct values** columns, we can reduce that query to only the visible fields. I think that would diminish the overall user experience, but that's a decision we can make to reduce the load on the query. 

For the information you see when you expand the rows (e.g. 10 categories, 2000-7000), there's another set of queries that we run to get the top terms for the keyword fields, distribution for numeric fields, etc. Those queries are more in-depth and heavier, and we fetch only for the visible fields in the table.

stratoula commented:
@qn895 gotcha! I think it is fine for now, especially with the async strategy. If users complain about the performance we can evaluate again

stratoula reviewed (approved):
Changes LGTM, thanx for addressing the comments! @jughosta can you continue with your review?

peteharverson commented:
> If we choose to disable sorting on the Documents (%) and Distinct values columns, we can reduce that query to only the visible fields. I think that would diminish the overall user experience, but that's a decision we can make to reduce the load on the query.

I would discourage you from disabling sorting on these fields - I commonly use this to find for examples the fields which occur in all / a tiny fraction of docs.

jughosta commented on src/plugins/discover/public/application/main/components/layout/discover_main_content.tsx:
```diff
@@ -60,13 +60,17 @@ export const DiscoverMainContent = ({
-      stateContainer.appState.update({ viewMode: mode });
+      stateContainer.appState.update(
+        // Resetting to default viewMode: undefined if document level
+        // else change for viewMode from undefined to DOCUMENT_LEVEL will triger unsaved changes badge
```
If I save a search with Field Statistics tab selected and then switch to Documents, I don't see the badge any more although it was there before the PR.

@kertal @davismcphee What behaviour is preferable?

kertal replied:
I'd say this badge should be displayed since the modified state is different to the persisted one, and if it's saved again, Documents instead of Field statistics are displayed when the saved search is loaded (and on a Dashboard).

davismcphee replied:
For data view mode the badge should be displayed when switching tabs since the selected tab is persisted to the saved search, so it represents a change in the state.

For ES|QL mode it depends on the intended behaviour. If we will support persisting the tab to the saved search, it _should_ should the badge. If we choose not to support persisting the tab, it _shouldn't_ show the badge. My personal preference is for consistency between modes, but I don't feel strongly about it so I'd defer to @stratoula on that one.

stratoula replied:
I agree with you Davis, if we don't persist the change then is not considered a state change. I will be completely honest here I am in the middle. On one hand I like consistency on the other hand I find it very weird to persist the fields statistics tab. Was it a feature request? I feel we should do something different here. But is def a business desicion.

My proposal is:

- merge this PR without the persistence. We are going to discuss it on Tue and if we decide in favor of it we can have a follow up PR
- wait till we have our sync on Tue, decide and merge after

It really depends on you @qn895 , I know how painful can be to keep a PR open for so long. I dont think also that is the end of the world if we merge the PR now and adjust in a week or so.

davismcphee replied:
I'm ok with merging with either behaviour since we're going to follow up on it after anyway, as long as the unsaved changes badge behaves correctly based on my previous comment.

I'm not sure if it was a feature request since it's been around longer than I have, but one reason I can think of for why it might work that way is that it's currently the only way to embed field statistics for a saved search within a Dashboard panel.

qn895 replied:
@davismcphee  @stratoula Thanks all for the reviews. To recap from our ES|QL sync, for this PR I will:
- Keep the state (documents vs field stats mode) saved in both non-ESQL and ESQL mode for consistency. So loading the saved search in the dashboard will persist the state when user saves the search. In the future, we can expand on the saved search experience to allow for finer grain control.
- Show the `unsaved changes` badge whenever the tab is changed, and update the functional tests to reflect that

davismcphee replied:
Yes that sounds right to me, thanks!

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ml-field-stats-esql-discover

qn895 committed:
Update view mode after merge

qn895 committed:
Add dashboard filter & query support

qn895 committed:
Clean up

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ml-field-stats-esql-discover

qn895 commented:
I've updated the PR to reflect the latest discussions:

- Keep the state (documents vs field stats mode) saved in both non-ESQL and ESQL mode for consistency. So loading the saved search in the dashboard will persist the state when user saves the search. In the future, we can expand on the saved search experience to allow for finer grain control.

    Saving a search in ES|QL while using Field statistics
    
    https://github.com/elastic/kibana/assets/43350163/14ab7394-1157-459b-a4cb-bf97d0e228c6
    
    ES|QL saved search in Dashboard:
    
    https://github.com/elastic/kibana/assets/43350163/601ab5c0-32e7-4e8b-b179-6840c91cf0c2
    

- Show the unsaved changes badge whenever the tab is changed

    https://github.com/elastic/kibana/assets/43350163/3ddb8c85-ae98-47e3-91fc-237687185369

- Remove usage of ES|QL API versioning

stratoula commented:
Looks great 🙌 @jughosta or @davismcphee can you review again pretty please?

davismcphee reviewed (commented):
@jughosta is out on PTO this week, so I'm picking up the remainder of the review from the Discover side. I left a couple of very minor comments, but overall the Data Discovery changes are looking good to me. @stratoula has a much better handle on if the generated ES|QL query is good, so I'll trust her review on that and focus on the behaviour in Discover.

There's just one minor issue I noticed related to tab switching and the unsaved changes badge in ES|QL mode. It seems like if open an ES|QL search and change the tab, the badge is triggered as expected, but if I revert the changes and continue switching tabs, the badge doesn't reappear until I refresh the page:

https://github.com/elastic/kibana/assets/25592674/9af3f89e-0dde-4097-a7dd-624dd3bdd84f

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ml-field-stats-esql-discover

qn895 committed:
Fix tests and outdated comments

qn895 commented:
Thanks @davismcphee  for the review. I've pushed an update that addressed all your comments. 

Unsaved changes badge:

https://github.com/elastic/kibana/assets/43350163/61fa752e-0884-44ec-83f0-92c9dde01c19

https://github.com/elastic/kibana/assets/43350163/a2094083-5dd3-49c7-a5c9-5c2292c64f49

davismcphee reviewed (approved):
Data Discovery changes LGTM 👍 Thanks for the great work on this!

I'm unfortunately still encountering the unsaved changes issue from my video, specifically in ES|QL saved searches, but I don't think it's significant enough to hold up this PR over, so I'm approving now and opened a followup issue for the bug here: #184624.

davismcphee committed:
Merge branch 'main' into ml-field-stats-esql-discover

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ml-field-stats-esql-discover

qn895 commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into ml-field-stats-esql-discover

qn895 committed:
Fix types
